File: .github/scripts/code-quality-ai/github_helpers.py
#!/usr/bin/env python3
"""GitHub API helpers for the code quality AI pipeline.

Authentication uses the GITHUB_TOKEN environment variable, which is
provided automatically when running inside a GitHub Actions workflow.
"""
from __future__ import annotations

import os
import re
from pathlib import Path

import requests

GITHUB_API = "https://api.github.com"
TIMEOUT_SECONDS = 30


def _auth_headers() -> dict[str, str]:
    token = os.environ.get("GITHUB_TOKEN")
    if not token:
        raise RuntimeError("GITHUB_TOKEN environment variable is required.")
    return {
        "Accept": "application/vnd.github+json",
        "Authorization": f"Bearer {token}",
        "X-GitHub-Api-Version": "2022-11-28",
    }


def open_issue(
    repo: str,
    title: str,
    body: str,
    labels: list[str] | None = None,
    assignees: list[str] | None = None,
) -> int:
    """Open a GitHub issue in `<owner>/<name>`. Returns the issue number."""
    payload: dict = {"title": title, "body": body}
    if labels:
        payload["labels"] = labels
    if assignees:
        payload["assignees"] = assignees
    resp = requests.post(
        f"{GITHUB_API}/repos/{repo}/issues",
        headers=_auth_headers(),
        json=payload,
        timeout=TIMEOUT_SECONDS,
    )
    resp.raise_for_status()
    return resp.json()["number"]


def open_pr(
    repo: str,
    title: str,
    body: str,
    head: str,
    base: str = "main",
    draft: bool = False,
) -> int:
    """Open a pull request in `<owner>/<name>`. Returns the PR number."""
    resp = requests.post(
        f"{GITHUB_API}/repos/{repo}/pulls",
        headers=_auth_headers(),
        json={
            "title": title,
            "body": body,
            "head": head,
            "base": base,
            "draft": draft,
        },
        timeout=TIMEOUT_SECONDS,
    )
    resp.raise_for_status()
    return resp.json()["number"]


def add_labels(repo: str, number: int, labels: list[str]) -> None:
    """Add labels to an issue or PR (PRs share the issues endpoint)."""
    resp = requests.post(
        f"{GITHUB_API}/repos/{repo}/issues/{number}/labels",
        headers=_auth_headers(),
        json={"labels": labels},
        timeout=TIMEOUT_SECONDS,
    )
    resp.raise_for_status()


def request_reviewers(
    repo: str,
    pr_number: int,
    reviewers: list[str] | None = None,
    team_reviewers: list[str] | None = None,
) -> None:
    """Request user and/or team reviewers on a PR."""
    payload: dict = {}
    if reviewers:
        payload["reviewers"] = reviewers
    if team_reviewers:
        payload["team_reviewers"] = team_reviewers
    if not payload:
        return
    resp = requests.post(
        f"{GITHUB_API}/repos/{repo}/pulls/{pr_number}/requested_reviewers",
        headers=_auth_headers(),
        json=payload,
        timeout=TIMEOUT_SECONDS,
    )
    resp.raise_for_status()


_CODEOWNERS_LOCATIONS = (
    "CODEOWNERS",
    ".github/CODEOWNERS",
    "docs/CODEOWNERS",
)


def _load_codeowners(repo_root: Path) -> list[tuple[str, list[str]]]:
    """Parse CODEOWNERS into ordered (pattern, owners) tuples."""
    for relative in _CODEOWNERS_LOCATIONS:
        path = repo_root / relative
        if path.is_file():
            entries: list[tuple[str, list[str]]] = []
            for raw_line in path.read_text().splitlines():
                line = raw_line.strip()
                if not line or line.startswith("#"):
                    continue
                pattern, *owners = line.split()
                if owners:
                    entries.append((pattern, owners))
            return entries
    return []


def _pattern_to_regex(pattern: str) -> re.Pattern[str]:
    anchored = pattern.startswith("/")
    if anchored:
        pattern = pattern.lstrip("/")
    body = re.escape(pattern).replace(r"\*\*", ".*").replace(r"\*", "[^/]*")
    if pattern.endswith("/"):
        body += ".*"
    if anchored:
        body = "^" + body
    return re.compile(body)


def lookup_codeowners(path: str, repo_root: Path | None = None) -> list[str]:
    """Return CODEOWNERS for a file path. The last matching rule wins."""
    repo_root = repo_root or Path.cwd()
    matched: list[str] = []
    for pattern, owners in _load_codeowners(repo_root):
        if _pattern_to_regex(pattern).search(path):
            matched = owners
    return matched

File: .github/scripts/code-quality-ai/github_helpers.js
#!/usr/bin/env node
/**
 * GitHub API helpers for the code quality AI pipeline.
 *
 * Authentication uses the GITHUB_TOKEN environment variable, which is
 * provided automatically when running inside a GitHub Actions workflow.
 */
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";

import { Octokit } from "@octokit/rest";

function octokit() {
  const token = process.env.GITHUB_TOKEN;
  if (!token) {
    throw new Error("GITHUB_TOKEN environment variable is required.");
  }
  return new Octokit({ auth: token });
}

function splitRepo(repo) {
  const [owner, name] = repo.split("/");
  if (!owner || !name) {
    throw new Error(`Invalid repo "${repo}", expected "owner/name".`);
  }
  return { owner, repo: name };
}

/**
 * Stable, search-friendly token derived from a finding_id. Embedded in
 * issue/PR titles so we can detect duplicates without needing to store
 * state across runs. The same finding_id always yields the same token.
 *
 * @param {string} findingId
 * @returns {string} Token of the form `ai-fid-<12 hex chars>`.
 */
export function findingIdToken(findingId) {
  const hash = crypto.createHash("sha256").update(findingId).digest("hex");
  return `ai-fid-${hash.slice(0, 12)}`;
}

/**
 * Search the repo for an existing issue or PR whose title contains the
 * given finding-id token. Returns the first match (any state — open or
 * closed) or null. PRs are issues at the API level, so a single search
 * covers both tracks.
 *
 * @param {string} repo "owner/name"
 * @param {string} token Token from `findingIdToken`.
 * @returns {Promise<{number: number, state: string, html_url: string, pull_request?: object} | null>}
 */
export async function findExistingByToken(repo, token) {
  const { owner, repo: name } = splitRepo(repo);
  const q = `repo:${owner}/${name} in:title "${token}"`;
  const resp = await octokit().search.issuesAndPullRequests({
    q,
    per_page: 5,
  });
  const items = resp.data.items ?? [];
  return items.length > 0 ? items[0] : null;
}

/**
 * Open a GitHub issue. Returns the issue number.
 *
 * @param {string} repo "owner/name"
 * @param {{title: string, body: string, labels?: string[], assignees?: string[]}} args
 * @returns {Promise<number>}
 */
export async function openIssue(repo, { title, body, labels, assignees }) {
  const { owner, repo: name } = splitRepo(repo);
  const resp = await octokit().issues.create({
    owner,
    repo: name,
    title,
    body,
    labels,
    assignees,
  });
  return resp.data.number;
}

/**
 * Open a pull request. Returns the PR number.
 *
 * @param {string} repo "owner/name"
 * @param {{title: string, body: string, head: string, base?: string, draft?: boolean}} args
 * @returns {Promise<number>}
 */
export async function openPr(
  repo,
  { title, body, head, base = "main", draft = false },
) {
  const { owner, repo: name } = splitRepo(repo);
  const resp = await octokit().pulls.create({
    owner,
    repo: name,
    title,
    body,
    head,
    base,
    draft,
  });
  return resp.data.number;
}

/**
 * Add labels to an issue or PR (PRs share the issues endpoint).
 *
 * @param {string} repo "owner/name"
 * @param {number} number Issue or PR number.
 * @param {string[]} labels
 */
export async function addLabels(repo, number, labels) {
  const { owner, repo: name } = splitRepo(repo);
  await octokit().issues.addLabels({
    owner,
    repo: name,
    issue_number: number,
    labels,
  });
}

/**
 * Request user and/or team reviewers on a PR. No-op if both lists are empty.
 *
 * @param {string} repo "owner/name"
 * @param {number} prNumber
 * @param {{reviewers?: string[], teamReviewers?: string[]}} [args]
 */
export async function requestReviewers(
  repo,
  prNumber,
  { reviewers, teamReviewers } = {},
) {
  if (!reviewers?.length && !teamReviewers?.length) return;
  const { owner, repo: name } = splitRepo(repo);
  await octokit().pulls.requestReviewers({
    owner,
    repo: name,
    pull_number: prNumber,
    reviewers,
    team_reviewers: teamReviewers,
  });
}

const CODEOWNERS_LOCATIONS = [
  "CODEOWNERS",
  ".github/CODEOWNERS",
  "docs/CODEOWNERS",
];

function loadCodeowners(repoRoot) {
  for (const rel of CODEOWNERS_LOCATIONS) {
    const p = path.join(repoRoot, rel);
    if (fs.existsSync(p)) {
      const entries = [];
      const lines = fs.readFileSync(p, "utf8").split("\n");
      for (const raw of lines) {
        const line = raw.trim();
        if (!line || line.startsWith("#")) continue;
        const [pattern, ...owners] = line.split(/\s+/);
        if (owners.length) entries.push([pattern, owners]);
      }
      return entries;
    }
  }
  return [];
}

function patternToRegex(pattern) {
  let p = pattern;
  const anchored = p.startsWith("/");
  if (anchored) p = p.replace(/^\/+/, "");
  // Escape regex specials including '*' and '?', then expand the escaped
  // glob forms. Escaping '*' first prevents the second replacement from
  // touching the '*' produced by '**' -> '.*'.
  let body = p.replace(/[.+^${}()|[\]\\*?]/g, "\\$&");
  body = body.replace(/\\\*\\\*/g, ".*").replace(/\\\*/g, "[^/]*");
  if (p.endsWith("/")) body += ".*";
  if (anchored) body = "^" + body;
  return new RegExp(body);
}

/**
 * Return CODEOWNERS for a file path. Last matching rule wins, mirroring
 * GitHub's CODEOWNERS evaluation semantics.
 *
 * @param {string} filePath
 * @param {string} [repoRoot=process.cwd()]
 * @returns {string[]}
 */
export function lookupCodeowners(filePath, repoRoot = process.cwd()) {
  let matched = [];
  for (const [pattern, owners] of loadCodeowners(repoRoot)) {
    if (patternToRegex(pattern).test(filePath)) {
      matched = owners;
    }
  }
  return matched;
}
